src/flag_gems/ops/randperm.py

```diff
@@ -0,0 +1,412 @@
+            ) * bins + bin_id
+            partial_prefix = tl.load(d_lookback + rd_lbk_offset, volatile=True)
+            while partial_prefix == 0:
+                partial_prefix = tl.load(d_lookback + rd_lbk_offset, volatile=True)
+            inc_sum += (partial_prefix & LOOKBACK_VALUE_MASK).to(tl.int32)
+            if partial_prefix & LOOKBACK_GLOBAL_MASK:
+                # break
+                bk = -1
+            else:
+                bk -= 1
+        global_counter = inc_sum | LOOKBACK_GLOBAL_MASK
+        tl.store(
+            d_lookback
+            + ((portion_id * passes + p) * max_tiles_per_portion + pid0) * bins
+            + bin_id,
+            global_counter,
+            cache_modifier=".wt",
+        )
+        inc_bucket_offset = prefix_offsets.to(tl.int64) + inc_sum.to(tl.int64)
+        if last_block and portion_id < num_portions - 1:
+            tl.store(
+                digit_hist + bin_offset + (portion_id + 1) * passes * (bins + 1),
+                inc_bucket_offset,
+            )
+        global_offsets = (
+            inc_bucket_offset - bin_of_bucket.to(tl.int64) + key_block_rank.to(tl.int64)
+        )
+        tl.store(key_out + global_offsets, key_data, mask=key_digit_mask)
+        tl.store(value_out + global_offsets, value_data, mask=key_digit_mask)
+
+
+# for parallelization, randomly shuffle the entire block rather than adjacent equal elements as pytorch cuda backend
+@libentry()
+@triton.jit(do_not_specialize=["philox_seed", "philox_offset"])
+def duplicate_keys_shuffle_kernel(
+    value_in, n_elements, philox_seed, philox_offset, BLOCK_SIZE: tl.constexpr
+):
+    pid0 = tl.program_id(0)
+    offset_range = tl.arange(0, BLOCK_SIZE)
+    value_offset = pid0.to(tl.int64) * BLOCK_SIZE + offset_range
+    value_mask = value_offset < n_elements
+    value_data = tl.load(value_in + value_offset, mask=value_mask)
+
+    philox_seed = philox_seed.to(tl.int64)
+    philox_offset = philox_offset.to(tl.int64)
+    c0 = (philox_offset & 0xFFFFFFFF).to(tl.uint32)
+    c1 = ((philox_offset >> 32) & 0xFFFFFFFF).to(tl.uint32)
+    i4 = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    c0 += i4
+    _O = c0 * 0
+    r0, _, _, _ = tl.philox(philox_seed, c0, c1, _O, _O)
+
+    _block_size = BLOCK_SIZE
+    r1 = r0 % _block_size.to(tl.uint32)
+    mask_val = _get_iinfo_val(tl.uint32, True)
+    r1 = tl.where(value_offset < n_elements, r1, mask_val)
+    _, sorted_chunk_index = argsort(r1, offset_range, 0, descending=False)
+    store_offset = pid0.to(tl.int64) * BLOCK_SIZE + sorted_chunk_index.to(tl.int64)
+    tl.store(value_in + store_offset, value_data, mask=store_offset < n_elements)
+
+
+def sort_by_key(key, value):
+    n_elements = key.numel()
+    if n_elements > 2 * 1024:
+        # radix method
+        BLOCK_SIZE = 1024
+        bits_per_pass = 4
+        bits_per_segment = 3
+        passes = triton.cdiv(key.element_size() * 8, bits_per_pass)
```

Comment:
Can we cut down the `passes` here by calculating the `start_bit` and `end_bit` based on the given `n_elements` to minimize iterations？I'm not sure if this approach will work

Reply:
Good suggestions! Adjust the  bits of key according to the given ```n_elements```, and thus ```passses``` may be cut down in most cases. The following picture shows the current flaggems perf vs pytorch cuda on A100:

```
Size    Torch Latency (ms)    Gems Latency (ms)    Gems Speedup
---------------------------------------------------------------
1024              0.027648              0.08192           0.338
6144              0.095232             0.288768            0.33
11264             0.103424              0.28672           0.361
16384             0.103424             0.287744           0.359
21504             0.099328             0.287744           0.345
26624             0.099328             0.287744           0.345
31744             0.099328             0.285696           0.348
36864             0.101376             0.284672           0.356
41984             0.104448             0.284672           0.367
47104             0.103424             0.287744           0.359
52224             0.106496              0.28672           0.371
57344             0.109568             0.288768           0.379
62464             0.106496             0.289792           0.367
67584              0.10752              0.34304           0.313
72704             0.111616              0.34304           0.325
77824              0.10752             0.345088           0.312
Operator randperm Performance Test (dtype=torch.int64, mode=cuda)
Size    Torch Latency (ms)    Gems Latency (ms)    Gems Speedup
---------------------------------------------------------------
1024                0.0256             0.123904           0.207
6144              0.075776             0.290816           0.261
11264             0.078848             0.290816           0.271
16384              0.08192             0.287744           0.285
21504             0.083968             0.288768           0.291
26624             0.086016             0.288768           0.298
31744              0.10752             0.288768           0.372
36864             0.108544             0.288768           0.376
41984             0.108544             0.288768           0.376
47104             0.109568             0.287744           0.381
52224             0.109568             0.288768           0.379
57344              0.11264             0.287744           0.391
62464             0.113664             0.287744           0.395
67584             0.114688             0.344064           0.333
72704             0.115712              0.34304           0.337
77824             0.115712             0.344064           0.336
```